sbi/samplers/vi/vi_divergence_optimizers.py

```diff
@@ -0,0 +1,710 @@
+                based on the 'reduced_variance' already present, but leads to an
+                unbiased estimate on this objective in contrast.
+
+        References:
+            [1] _Doubly Reparameterized Gradient Estimators for Monte Carlo Objectives_,
+                George Tucker, Dieterich Lawson, Shixiang Gu, Chris J. Maddison, 2018,
+                https://arxiv.org/abs/1810.04152.
+
+
+        """
+        super().__init__(*args, **kwargs)
+        self.K = K
+        self.loss_name = "iwelbo"
+        self.eps = 1e-7
+        self.dreg = dreg
+        self.HYPER_PARAMETERS += ["K", "dreg"]
+        if dreg:
+            self.reduce_variance = True
+
+    def loss_rsample(self, x_obs: Tensor) -> Tuple[Tensor, Tensor]:
+        """Computes the IWELBO loss.
+
+        Args:
+            x_obs: Observation
+
+        Returns:
+            surrogate_loss: Loss to differentiate.
+            loss: Loss to diplay.
+
+        """
+        elbo_particles = self.generate_elbo_particles(x_obs, self.n_particles * self.K)
+        elbo_particles = elbo_particles.reshape(self.n_particles, self.K)
+        weights = self.get_importance_weight(elbo_particles.clone().detach())
+        surrogate_loss = -(weights * elbo_particles).sum(-1).mean(0)
+        loss = -torch.mean(torch.exp(elbo_particles) + 1e-20, -1).log().mean()
+        return surrogate_loss, loss.clone().detach()
+
+    def get_importance_weight(self, elbo_particles: Tensor) -> Tensor:
+        """Generates importance weights in a numerically stable fashion.
+
+        Args:
+            elbo_particles: ELements contain log potential - log q.
+
+        Returns:
+            Tensor: Normalized importance weights.
+
+        """
+        logweights = elbo_particles
+        normalized_weights = torch.exp(
+            logweights - torch.logsumexp(logweights, -1).unsqueeze(-1)
+        )
+        if self.dreg:
+            normalized_weights = normalized_weights ** 2
+        return normalized_weights
+
+
+@register_VI_method(name="fKL")
+class ForwardKLOptimizer(DivergenceOptimizer):
+    """This learns the variational posterior by minimizeing the forward KL divergence
+    i.e. D_KL(p(theta|x_o)|| q(theta)). The typically necessary samples from the
+    posterior are not required by using importance sampling tenchniques.
+
+    NOTE: Whereas in the previous cases n_particles mainly reduces the variance, in this
+    case it also decrease the bias. Using n_particles=1 is not usefull.
+
+    References:
+        [1] _Variational Refinement for Importance Sampling Using the Forward
+            Kullback-Leibler Divergence_, Ghassen Jerfel, Serena Wang, Clara Fannjiang,
+            Katherine A. Heller, Yian Ma, Michael I. Jordan, 2021,
+            https://arxiv.org/abs/2106.15980.
+    """
+
+    def __init__(
+        self,
+        *args,
+        proposal: str = "q",
+        is_method: str = "identity",
```

Comment:
Did you use anything other than `"identity"` in the paper? If not, I'd say we remove these options (purely to reduce the amount of code on has to test and maintain)

Reply:
I only used identity, so yeah we can remove it. We may can make the weight_f: Callable such that the user can pass it if he wants.